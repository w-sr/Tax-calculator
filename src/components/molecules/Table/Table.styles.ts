import styled, { css } from "styled-components";

export const Table = styled.table`
  border-collapse: collapse;
  border: none;
`;

export const TableHeaderCell = styled.th`
  ${({ theme }) => css`
    background-color: ${theme.colors.greyDark};
    padding: 12px;
    text-align: left;
    font-size: ${theme.font.small.size};
    line-height: ${theme.font.small.lineHeight};
    font-weight: ${theme.font.small.weight};
    color: ${theme.colors.black};
    &:first-child {
      border-top-left-radius: 12px;
    }
    &:last-child {
      border-top-right-radius: 12px;
    }
  `}
`;

export const TableRowItem = styled.tr`
  ${({ theme }) => css`
    cursor: auto;
    &:nth-child(odd) {
      background-color: ${theme.colors.greyWhite};
    }
    &:last-child {
      border-bottom-left-radius: 12px;
      border-bottom-right-radius: 12px;
    }
  `}
`;

export const TableCell = styled.td`
  ${({ theme }) => css`
    padding: 12px;
    font-size: ${theme.font.small.size};
    line-height: ${theme.font.small.lineHeight};
    font-weight: ${theme.font.small.weight};
    color: ${theme.colors.greyWeb};
  `}
`;
