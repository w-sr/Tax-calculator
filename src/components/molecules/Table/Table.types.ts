export interface IColumnType<T> {
  key: string;
  title: string;
  width?: number;
  render?: (column: IColumnType<T>, item: T) => void;
}
