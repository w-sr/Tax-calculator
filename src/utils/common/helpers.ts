const buildPercent = (decimal?: number) =>
  new Intl.NumberFormat("en-US", {
    style: "percent",
    minimumFractionDigits: decimal ?? 2,
  });

/**
 *
 * @param value
 * @param decimal
 * @returns value with percentage
 * @example 0.20 => 20%
 */
export const percentValue = (
  value: number | undefined,
  decimal?: number
): string => {
  if (value === undefined) return "N/A";
  const percent = buildPercent(decimal);
  return percent.format(value);
};

/**
 *
 * @param value
 * @param decimal
 * @returns value with currency
 * @example 20 => $20
 */
export const moneyValue = (
  value: number | undefined,
  decimal?: number
): string => {
  const money = new Intl.NumberFormat("en-US", {
    style: "currency",
    currency: "USD",
    currencyDisplay: "narrowSymbol",
    minimumFractionDigits: decimal ?? 2,
  });
  return money.format(value ?? 0);
};

/**
 *
 * @param min
 * @param max
 * @returns formatted bracket
 * @example min: 0, max: 2000 => "$2000 or less"
 *          min: 2000, max: 4000 => "$2000 to $4000"
 *          min: 4000 => "More than $4000"
 */
export const bracketFormat = (min: number, max?: number): string => {
  if (min === 0) return moneyValue(max, 0) + " or less";
  if (max) return moneyValue(min, 0) + " to " + moneyValue(max, 0);
  else return "More than " + moneyValue(min, 0);
};
