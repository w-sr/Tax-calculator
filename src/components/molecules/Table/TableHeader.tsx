import { IColumnType } from "./Table.types";
import * as S from "./Table.styles";

interface Props<T> {
  columns: IColumnType<T>[];
}

export function TableHeader<T>({ columns }: Props<T>): JSX.Element {
  return (
    <tr>
      {columns.map((column, columnIndex) => (
        <S.TableHeaderCell
          key={`table-head-cell-${columnIndex}`}
          style={{ width: column.width }}
        >
          {column.title}
        </S.TableHeaderCell>
      ))}
    </tr>
  );
}
