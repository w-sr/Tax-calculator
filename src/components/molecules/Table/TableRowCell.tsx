import get from "lodash.get";
import { IColumnType } from "./Table.types";
import * as S from "./Table.styles";

interface Props<T> {
  item: T;
  column: IColumnType<T>;
}

export function TableRowCell<T>({ item, column }: Props<T>): JSX.Element {
  const value = get(item, column.key);
  return (
    <S.TableCell>
      {column.render ? column.render(column, item) : value}
    </S.TableCell>
  );
}
