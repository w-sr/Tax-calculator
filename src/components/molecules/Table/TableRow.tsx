import { IColumnType } from "./Table.types";
import { TableRowCell } from "./TableRowCell";
import * as S from "./Table.styles";

interface Props<T> {
  data: T[];
  columns: IColumnType<T>[];
}

export function TableRow<T>({ data, columns }: Props<T>): JSX.Element {
  return (
    <>
      {data.map((item, itemIndex) => (
        <S.TableRowItem key={`table-body-${itemIndex}`}>
          {columns.map((column, columnIndex) => (
            <TableRowCell
              key={`table-row-cell-${columnIndex}`}
              item={item}
              column={column}
            />
          ))}
        </S.TableRowItem>
      ))}
    </>
  );
}
