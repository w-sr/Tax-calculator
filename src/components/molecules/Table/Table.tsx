import { IColumnType } from "./Table.types";
import { TableHeader } from "./TableHeader";
import { TableRow } from "./TableRow";

interface Props<T> {
  data: T[];
  columns: IColumnType<T>[];
}

export function Table<T>({ data, columns }: Props<T>): JSX.Element {
  return (
    <table>
      <thead>
        <TableHeader columns={columns} />
      </thead>
      <tbody>
        <TableRow data={data} columns={columns} />
      </tbody>
    </table>
  );
}
