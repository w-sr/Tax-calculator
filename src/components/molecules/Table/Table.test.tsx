import { renderWithTheme } from "../../../utils/tests/helpers";
import { Table } from "./Table";

interface IData {
  bracket: string;
  rate: string;
  tax: string;
}
const columns = [
  {
    key: "bracket",
    title: "Tax Bracket",
    width: 150,
  },
  {
    key: "rate",
    title: "Tax Rate",
    width: 150,
  },
  {
    key: "tax",
    title: "Tax Per Band",
    width: 150,
  },
];

const data: IData[] = [];

describe("<Table />", () => {
  it("should match the snapshot", () => {
    const { container } = renderWithTheme(
      <Table columns={columns} data={data} />
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
