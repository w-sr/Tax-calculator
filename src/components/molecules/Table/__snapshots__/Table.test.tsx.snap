// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> should match the snapshot 1`] = `
<table>
  <thead>
    <tr>
      <th
        class="sc-gueYoa klxFfn"
        style="width: 150px;"
      >
        Tax Bracket
      </th>
      <th
        class="sc-gueYoa klxFfn"
        style="width: 150px;"
      >
        Tax Rate
      </th>
      <th
        class="sc-gueYoa klxFfn"
        style="width: 150px;"
      >
        Tax Per Band
      </th>
    </tr>
  </thead>
  <tbody />
</table>
`;
